src/cargo/core/compiler/unit_dependencies.rs

```diff
@@ -857,3 +1121,27 @@ impl<'a, 'cfg> State<'a, 'cfg> {
             .collect()
     }
 }
+
+fn non_custom_and_non_transitive_deps(unit: &Unit, dep: &Dependency) -> bool {
```

Comment:
Indeed, `master` contained a refactoring which I effectively ended up reverting during the rebase, as the 'easiest' way for me to avoid breakage. Now it's time to clean it up and figure out what's really going on. 

The test case above is seemingly represented by `allow_artifact_and_non_artifact_dependency_to_same_crate()`, or so it seems, so I didn't add this as separate test while trying to make the necessary adjustments to return to the filter handling introduced in `master`.

It is indeed the case that [`any`](https://github.com/rust-lang/cargo/blob/c0e6abe384c2c6282bdd631e2f2a3b092043e6c6/src/cargo/core/compiler/unit_dependencies.rs#L1118) is subtle here and caused me to have to apply the filter multiple times to actually make it stick. There is probably a very good reason to have `any` there, and I will see how I can make it work.

Reply:
I took a long look and believe it can't be inline due to the need for a different filter [when obtaining artifact deps for the build scripts run unit](https://github.com/rust-lang/cargo/blob/9ebc4e156f0efa60751ae2199cd70dc71b90f769/src/cargo/core/compiler/unit_dependencies.rs#L522).

What I ended up doing in 9ebc4e156f0efa60751ae2199cd70dc71b90f769 is to provide a version with and without filter, making `state.deps(…)` similar in signature as in `master`.

I struggled a lot with [the `.expect("artifact")` line](https://github.com/rust-lang/cargo/blob/9ebc4e156f0efa60751ae2199cd70dc71b90f769/src/cargo/core/compiler/unit_dependencies.rs#L538) just to realize that indeed only artifacts can be present there due to the prior `.is_build()` filter. Currently it's not allowed to have a renamed dependency to the same package (and thus package id), but I believe this `expect()` could actually be encountered with the `multidep` toggle in RFC-3176. There I left a note to add a test for that which should motivate a fix.